pkg/controllers/user/logging/clusterlogging.go

```diff
@@ -60,48 +56,50 @@ func registerClusterLogging(cluster *config.UserContext) {
 		rolebindings:         cluster.RBAC.RoleBindings(loggingconfig.LoggingNamespace),
 		clusterRoleBindings:  cluster.RBAC.ClusterRoleBindings(loggingconfig.LoggingNamespace),
 		clusterLister:        cluster.Management.Management.Clusters("").Controller().Lister(),
+		nodeLister:           cluster.Management.Management.Nodes("").Controller().Lister(),
 		clusterName:          cluster.ClusterName,
 	}
-	clusterloggingClient.AddClusterScopedHandler("cluster-logging-controller", cluster.ClusterName, syncer.Sync)
+	clusterloggingClient.AddClusterScopedLifecycle("cluster-logging-controller", cluster.ClusterName, lifecycle)
 }
 
-func (c *ClusterLoggingSyncer) Sync(key string, obj *v3.ClusterLogging) error {
-	//clean up
-	if obj == nil || obj.DeletionTimestamp != nil {
-		return utils.CleanResource(c.namespaces, c.clusterLoggingLister, c.projectLoggingLister)
-	}
+func (c *ClusterLoggingLifecycle) Create(obj *v3.ClusterLogging) (*v3.ClusterLogging, error) {
+	newObj, err := v3.LoggingConditionProvisioned.DoUntilTrue(obj, func() (runtime.Object, error) {
+		return obj, c.provision(obj)
+	})
+	return newObj.(*v3.ClusterLogging), err
+}
 
-	if err := utils.IniteNamespace(c.namespaces); err != nil {
-		return err
-	}
-	if err := utils.InitConfigMap(c.configmaps); err != nil {
-		return err
+func (c *ClusterLoggingLifecycle) Updated(obj *v3.ClusterLogging) (*v3.ClusterLogging, error) {
+	newObj, err := v3.LoggingConditionProvisioned.DoUntilTrue(obj, func() (runtime.Object, error) {
+		return obj, c.provision(obj)
+	})
+
+	if err != nil {
+		return newObj.(*v3.ClusterLogging), err
 	}
 
-	//embedded
-	if utils.GetClusterTarget(obj.Spec) == "embedded" {
-		if err := utils.CreateOrUpdateEmbeddedTarget(c.deployments, c.serviceAccounts, c.services, c.roles, c.rolebindings, loggingconfig.LoggingNamespace, obj); err != nil {
-			return err
+	newObj, err = v3.LoggingConditionUpdated.Do(newObj, func() (runtime.Object, error) {
+		if err := utils.UpdateEmbeddedEndpoint(c.podLister, c.serviceLister, c.clusterLoggings, c.nodeLister, c.clusterName); err != nil {
```

Comment:
If the resource is not enough and the ES is not schedulable, will UpdateEmbeddedEndpoint return the error?

Reply:
And Update func only get called when the CRD is updated.  How UI get to know the progress of the ES deployment? maybe we should do something like cluster provisioning.